nsxt/resource_nsxt_host_upgrade_group_assignment.go

```diff
@@ -0,0 +1,197 @@
+/* Copyright © 2024 Broadcom, Inc. All Rights Reserved.
+   SPDX-License-Identifier: MPL-2.0 */
+
+package nsxt
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/vmware/vsphere-automation-sdk-go/lib/vapi/std/errors"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/bindings"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/protocol/client"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/model"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/upgrade"
+	"golang.org/x/exp/slices"
+)
+
+const hostUpgradeUnitDefaultGroup = "Group 1 for ESXI"
+
+func resourceNsxtHostUpgradeGroupAssignment() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceNsxtHostUpgradeGroupAssignmentCreate,
+		Read:   resourceNsxtHostUpgradeGroupAssignmentRead,
+		Update: resourceNsxtHostUpgradeGroupAssignmentUpdate,
+		Delete: resourceNsxtHostUpgradeGroupAssignmentDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceNsxtHostUpgradeGroupAssignmentImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"host_id": {
+				Type:        schema.TypeString,
+				Description: "The ID of the ESXi host that this group assigns to",
+				Required:    true,
+				ForceNew:    true,
+			},
+			"host_upgrade_group_id": {
+				Type:        schema.TypeString,
+				Description: "The ID of the ESXi host group that the host is assigned to",
+				Required:    true,
+			},
+		},
+	}
+}
+
+func getUnitIDsFromUnits(units []model.UpgradeUnit) []string {
+	var unitIDs []string
+
+	for _, unit := range units {
+		unitIDs = append(unitIDs, *unit.Id)
+	}
+	return unitIDs
+}
+
+func addHostToGroup(connector client.Connector, groupID, hostID string, isCreate bool) error {
+	client := upgrade.NewUpgradeUnitGroupsClient(connector)
+
+	group, err := client.Get(groupID, nil)
+	if err != nil {
+		return err
+	}
+
+	hostIDs := getUnitIDsFromUnits(group.UpgradeUnits)
+	if slices.Contains(hostIDs, hostID) {
+		return fmt.Errorf("host %s already exists in group %s", hostID, groupID)
+	}
+	group.UpgradeUnits = append(group.UpgradeUnits, model.UpgradeUnit{Id: &hostID})
+	_, err = client.Update(groupID, group)
```

Comment:
I feel simultaneous updates might cause errors here, so we might need to wrap this code (including `Get`) in retry. Auto-retry that we have today will not re-read the revision, so Update would fail again if revision got updated. What do you think?

Reply:
Makes a lot of sense. Do we have any example for such retry mechanism in the code?
BTW it seems like the `_revision` attribute is ignored by NSX - I'll open a bug for this.

Reply:
I think that I've found something.